docs/introduction/context-system.md

```diff
@@ -0,0 +1,170 @@
+
+class MyService
+{
+    public function __construct(
+        private readonly ContextResolverInterface $contextResolver,
+    ) {
+    }
+
+    public function doSomething(): void
+    {
+        if ($this->contextResolver->isCurrentContext(AdminContext::class)) {
+            // This code runs only in admin
+            $this->doAdminSpecificLogic();
+        }
+
+        if ($this->contextResolver->isCurrentContext(ConsoleContext::class)) {
+            // This code runs only in console commands
+            $this->doConsoleSpecificLogic();
+        }
+    }
+}
+```
+
+## Creating Custom Contexts
+
+### 1. Create Context Class
+
+Create a new context by extending `AbstractContext`:
+
+```php
+<?php
+
+namespace App\Component\Context;
+
+use Shopsys\FrameworkBundle\Component\Context\AbstractContext;
+
+final class ApiContext extends AbstractContext
+{
+    public function matches(): bool
+    {
+        // Your matching logic here
+        return $this->request->getPathInfo() === '/api';
+    }
+
+    public function getDescription(): string
+    {
+        return 'REST API requests';
+    }
+}
+```
+
+### 2. Context with Dependencies
+
+You can create contexts that depend on other contexts:
+
+```php
+<?php
+
+namespace App\Component\Context;
+
+use Shopsys\FrameworkBundle\Component\Context\AbstractContext;
+use Shopsys\FrameworkBundle\Component\Context\AdminContext;
+
+final class AdminApiContext extends AbstractContext
+{
+    public function getRequiredContexts(): array
+    {
+        // This context only matches if AdminContext also matches
+        return [AdminContext::class];
+    }
+
+    public function matches(): bool
+    {
+        // Your implementation
+    }
+
+    public function getDescription(): string
+    {
+        return 'Matches Admin API requests';
+    }
+}
+```
+
+## Context Dependencies
+
+### How Dependencies Work
+
+Context dependencies ensure that a context only matches when all its required contexts also match. This creates a hierarchical system:
+
+```
+ConsoleContext (base CLI context)
+├── ConsumerContext (message processing)
+└── CronContext (scheduled jobs)
+    └── ImportContext (custom import commands)
+```
+
+## Best Practices
+
+### 1. Keep Contexts Simple
+
+All contexts should have a single, clear purpose. Avoid complex logic that could change frequently, as this can lead to maintenance issues.
+Also matches() method should be performant and not do any heavy operations like database queries or complex calculations as all contexts are checked on every request.
+
+```php
+// Good: Simple, focused logic
+public function matches(): bool
+{
+    return $this->request->getPathInfo() === '/api';
+}
+
+// Avoid: Complex logic that could change
+public function matches(): bool
+{
+    return $this->request->getPathInfo() === '/api'
+        && $this->request->getMethod() === 'POST'
+        && $this->request->headers->get('Content-Type') === 'application/json';
+}
```

Comment:
I can imagine something like this is needed one day :smile: I would rather put some DB queries into the bad example

Reply:
I changed that example